Add support for indented HEREDOC terminators

jen20 opened the pull request:
This PR adds support for the style of HEREDOC often used in Ruby which allows the terminating marker to be indented if the HEREDOC is started with the sequence "<<-" rather than the usual "<<". This allows users to express documents with more natural indentation:

```
resource "template_file" "test" {
    template = <<-HEREDOC
        First Line
        Second Line
    HEREDOC
}
```

Note that this does not attempt to add any semantics around removing hanging indent from the actual text of the document, so extra indentation would still be present. We could make use of the canonical style for HCL herre to remove the hanging indent in the style of Ruby which would probably be more predictable for users.

Fixes https://github.com/hashicorp/hcl/issues/9

cc @phinze, @mitchellh

phinze commented:
Haven't reviewed the code yet, but +1 from me on the fuctionality, I have definitely missed this functionality from our HEREDOC implementation thus far.

I think removing the hanging indent is worth doing - it's a nicer experience for config authors.

mitchellh commented:
I super want this. I started adding this awhile back but stopped due to complications primarily with the printer package. I see you haven't gotten there yet but I predict it will be fairly annoying. 

I agree that if we do this I think we should automatically remove the equivalent indent from the text. That is what makes this really very useful. The rules around this are somewhat odd, but I think it should be:

* If all text has whitespace characters at least equal to the whitespace before the ending HEREDOC, trim it.

* If any text has whitespace less than the whitespace before the ending HEREDOC (i.e. the text itself is not indented), then ignore all lines, and don't trim _any_ whitespace.

I know other languages that do this, instead of going with above I would just see their rules and copy, probably. No need for us to reinvent the wheel.

jen20 commented:
My plan for the whitespace was to copy what Ruby does - the rules you mention do sound sensible however. I haven't yet looked at the printer - but it should be valid in the first instance to output a regular unindented heredoc as semantically equivalent? When I come back to this (after Terraform 0.6.12) I'll do a roundup of popular implementations of this and post in here.

mitchellh commented:
That sounds like a great idea @jen20.

If I remember, the most clever language around heredocs that I actually liked was Scala. I think. Scala itself is a really complicated language but I remember the heredoc rules working as I expected and making the code look pretty. It may have been something else though...

jen20 commented:
I've implemented a first pass at this using the semantics explained in the commit message and added some tests around expected behaviour. I think it matches the description @mitchellh gave in [this earlier comment](https://github.com/hashicorp/hcl/pull/91#issuecomment-187338389). Effectively:

- If your heredoc starts with `<<`, results are as before
- If your heredoc starts with `<<-`, and every line has the same prefix (tabs/spaces in the same order) as the terminating marker line, the prefix is removed from each line
- If your heredoc starts with `<<-`, but not every line has the prefix, the heredoc is returned in the same manner as if it specified with `<<`, but with the whitespace leading the marker removed.

This requires some testing on Windows - I am not sure of the effect of line endings.

phinze commented:
Behavior sounds good. Code looks reasonable to me. Windows testing does look like it will be importangt though.

sethvargo commented:
@jen20 @phinze one thing I just learned in bash land that might be helpful in Terraform too is the following. If you quote the delimiter, no interpolation takes place inside the heredoc. This is useful if you want to put bash in your heredoc. In the case of Terraform, I can envision a scenario where I want to dynamically write a configuration file that uses a similar syntax to Terraform:

```hcl
inline = <<"EOF"
${this.is.plain-text} and not interpreted as Terraform syntax
EOF
```

mitchellh commented:
@sethvargo I'd say thats a little too non-obvious. The `<<` and `<<-` are well mimicked syntactical notions. We also don't have any other way to make a string "raw" so I think we can skip that for me.

Windows would be important. As long as it behaves like the old heredocs then we're good. We should probably add tests for that.

jen20 commented:
Right now it does behave like a regular here doc on Windows, but the tests fail because autocrlf screws up the expectation. I'll write some tests with the windows line endings hard coded in go code rather than in a separate HCL file.

jen20 commented:
I think this is good to go now - we have green CI on both AppVeyor (Windows) and Travis.

phinze commented:
LGTM
